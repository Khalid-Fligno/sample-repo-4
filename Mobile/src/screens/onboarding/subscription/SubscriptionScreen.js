import React, { useEffect, useState } from "react";
import {
  View,
  ScrollView,
  Text,
  NativeModules,
  Alert,
  Linking,
  ImageBackground,
  Platform,
} from "react-native";
import * as Haptics from "expo-haptics";
import appsFlyer from "react-native-appsflyer";
import { auth, db } from "../../../config/firebase";
import {
  discountedIdentifiers,
  identifiers,
  compareProducts,
  validateReceiptProduction,
  validateReceiptSandbox,
  compareInApp,
} from "../../../config/apple";
import SubscriptionTile from "../../../components/Onboarding/SubscriptionTile";
import NativeLoader from "../../../components/Shared/NativeLoader";
import Icon from "../../../components/Shared/Icon";
import colors from "../../../styles/colors";
import RNIap, {
  finishTransaction,
  purchaseErrorListener,
  purchaseUpdatedListener,
} from "react-native-iap";
import CustomBtn from "../../../components/Shared/CustomBtn";
import { containerPadding } from "../../../styles/globalStyles";
import { ONBOARDINGIMG } from "../../../library/images/onBoardingImg/onBoardingImg";
import { styles } from "./style";
import { useStorage } from '../../../hook/storage'

export const SubscriptionScreen = ({ navigation }) => {

  const [loading, setLoading] = useState(false)
  const [products, setProducts] = useState(undefined)
  const [discountedProducts, setDiscountedProducts] = useState(undefined)
  const [selectedIndex, setSelectedIndex] = useState(0)

  const InAppUtils = NativeModules;
  const specialOffer = navigation.getParam("specialOffer", undefined)
  const productTitleMapIOS = {
    0: "Yearly ",
    1: "Monthly ",
  };
  const productTitleMapAndroid = {
    0: "Monthly ",
    1: "Yearly ",
  };
  const subscriptionPeriodMap = {
    "com.fitazfk.fitazfkapp.sub.fullaccess.monthly.discount": "month",
    "com.fitazfk.fitazfkapp.sub.fullaccess.yearly.discounted": "year",
    "com.fitazfk.fitazfkapp.sub.fullaccess.monthly.foundation": "month",
    "com.fitazfk.fitazfkapp.sub.fullaccess.yearly.foundation": "year",
    "com.fitazfk.fitazfkapp.sub.fullaccess.monthly": "month",
    "com.fitazfk.fitazfkapp.sub.fullaccess.yearly": "year",
  };
  const andriodSubscriptionTitleMap = {
    0: "year",
    1: "month",
  };

  const openLink = (url) => {
    Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
    Linking.openURL(url);
  };

  const logout = () => {
    try {
      useStorage.removeItem("uid");
      auth.signOut();
      navigation.navigate("Auth");
    } catch (err) {
      console.log("Error logging out: ", err)
    }
  };

  const convertToProduct = (products) => {
    const convertedProds = products.map((product) => {
      return {
        identifier: product.productId,
        primary: product.productId,
        priceString: product.localizedPrice,
        price: Number(product.price),
        currencyCode: product.currency,
      };
    });
    return convertedProds;
  };

  const androidSubscriptions = async () => {
    purchaseUpdateSubscription = purchaseUpdatedListener(
      async (purchase) => {
        const receipt = purchase.transactionReceipt;

        if (receipt) {
          try {
            await finishTransaction(purchase)
              .catch(() => setLoading(false));

            if (Platform.OS === "android") {
              handleAndroidPayment(purchase);
            }
          } catch (err) {
            Alert.alert("purchase error", error.message);
          }
        }
      }
    );

    purchaseErrorSubscription = purchaseErrorListener(
      (error) => {
        Alert.alert("purchase error", error.message);
      }
    );
  };

  const validate = async (receiptData) => {
    const validationData = await validateReceiptProduction(receiptData).catch(
      async (error) => {
        if (error.redirect) {
          const validationDataSandbox = await validateReceiptSandbox(
            receiptData
          );
          return validationDataSandbox;
        }
        return undefined;
      }
    );
    return validationData;
  };

  const iOSPurchaseProductIdentifier = async (
    productIdentifier,
    productPrice,
    productCurrencyCode
  ) => {
    InAppUtils.canMakePayments((canMakePayments) => {
      if (!canMakePayments) {
        setLoading(false)
        Alert.alert(
          "Not Allowed",
          "This device is not allowed to make purchases. Please check restrictions on device"
        );
        return;
      }
      InAppUtils.purchaseProduct(productIdentifier, async (error, response) => {
        if (error) {
          setLoading(false)
          Alert.alert("Purchase cancelled");
          return;
        }

        if (response && response.productIdentifier) {
          const validationData = await validate(
            response.transactionReceipt
          );
          if (validationData === undefined) {
            setLoading(false)
            Alert.alert("Receipt validation error");
            return;
          }
          const sortedInApp = validationData.receipt.in_app
            .slice()
            .sort(compareInApp);
          const isValid =
            sortedInApp[0] && sortedInApp[0].expires_date_ms > Date.now();
          if (isValid === true) {
            const uid = await useStorage.getItem("uid");
            const userRef = db.collection("users").doc(uid);
            const data = {
              subscriptionInfo: {
                receipt: response.transactionReceipt,
                expiry: sortedInApp[0].expires_date_ms,
                originalTransactionId: sortedInApp[0].original_transaction_id,
                originalPurchaseDate: sortedInApp[0].original_purchase_date_ms,
                productId: sortedInApp[0].product_id,
                platform: Platform.OS,
              },
            };
            await userRef.set(data, { merge: true });
            const eventName = "af_start_trial";
            const eventValues = {
              af_price: productPrice,
              af_currency: productCurrencyCode,
            };
            appsFlyer.trackEvent(eventName, eventValues);
            userRef.get().then(async (doc) => {
              setLoading(false)
              if (await doc.data().onboarded) {
                navigation.navigate("App");
              } else {
                navigation.navigate("Onboarding1", {
                  name: navigation.getParam("name", null),
                });
              }
            });
          } else if (isValid === false) {
            setLoading(false)
            Alert.alert("Purchase Unsuccessful");
          } else {
            setLoading(false)
            Alert.alert("Something went wrong", `${isValid.message}`);
          }
        }
      });
    });
  };

  const purchaseProduct = async (
    productIdentifier,
    productPrice,
    productCurrencyCode
  ) => {
    setLoading(true)
    Haptics.selectionAsync();
    if (productIdentifier === undefined) {
      Alert.alert("No subscription selected");
      return;
    }
    Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
    if (Platform.OS === "ios") {
      await iOSPurchaseProductIdentifier(
        productIdentifier,
        productPrice,
        productCurrencyCode
      );
    } else if (Platform.OS === "android") {
      RNIap.requestSubscription(productIdentifier);
    }
  };

  const iOSPurchaseDiscountedProductIdentifier = async (
    productIdentifier,
    productPrice,
    productCurrencyCode
  ) => {
    InAppUtils.canMakePayments((canMakePayments) => {
      if (!canMakePayments) {
        setLoading(false)
        Alert.alert(
          "Not Allowed",
          "This device is not allowed to make purchases. Please check restrictions on device"
        );
        return;
      }
      InAppUtils.loadProducts(discountedIdentifiers, (loadError) => {
        if (loadError) {
          setLoading(false)
          Alert.alert(
            "Unable to connect to the App Store",
            "Please try again later"
          );
        }
        InAppUtils.purchaseProduct(
          productIdentifier,
          async (error, response) => {
            if (error) {
              setLoading(false)
              Alert.alert("Purchase cancelled");
              return;
            }
            if (response && response.productIdentifier) {
              const validationData = await validate(
                response.transactionReceipt
              );
              if (validationData === undefined) {
                setLoading(false)
                Alert.alert("Receipt validation error");
                return;
              }
              const sortedInApp = validationData.receipt.in_app
                .slice()
                .sort(compareInApp);
              const isValid =
                sortedInApp[0] && sortedInApp[0].expires_date_ms > Date.now();
              if (isValid === true) {
                const uid = await AsyncStorage.getItem("uid");
                const userRef = db.collection("users").doc(uid);
                const data = {
                  subscriptionInfo: {
                    receipt: response.transactionReceipt,
                    expiry: sortedInApp[0].expires_date_ms,
                    originalTransactionId:
                      sortedInApp[0].original_transaction_id,
                    originalPurchaseDate:
                      sortedInApp[0].original_purchase_date_ms,
                    productId: sortedInApp[0].product_id,
                    platform: Platform.OS,
                  },
                };
                await userRef.set(data, { merge: true });
                // Appsflyer event tracking - Start Free Trial
                const eventName = "af_start_trial";
                const eventValues = {
                  af_price: productPrice,
                  af_currency: productCurrencyCode,
                };
                appsFlyer.trackEvent(eventName, eventValues);
                userRef.get().then(async (doc) => {
                  setLoading(false)
                  if (await doc.data().onboarded) {
                    navigation.navigate("App");
                  } else {
                    navigation.navigate("Onboarding1", {
                      name: navigation.getParam("name", null),
                    });
                  }
                });
              } else if (isValid === false) {
                setLoading(false)
                Alert.alert("Purchase Unsuccessful");
              } else {
                setLoading(false)
                Alert.alert("Something went wrong", `${isValid.message}`);
              }
            }
          }
        );
      });
    });
  };

  const retryLoadProducts = () => {
    setLoading(true)
    if (Platform.OS === "ios") {
      retryLoadiOSProducts();
    } else if (Platform.OS === "android") {
      loadAndroidProducts();
    }
  };

  const retryLoadiOSProducts = () => {
    InAppUtils.loadProducts(identifiers, (error, products) => {
      if (error) {
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later"
        );
      } else if (products.length !== 2) {
        // IAP products not retrieved (App Store server down, etc.)
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later",
          [
            {
              text: "Cancel",
              style: "cancel",
            },
            {
              text: "Try Again",
              onPress: () => loadiOSProducts(),
            },
          ],
          { cancelable: false }
        );
      } else {
        const sortedProducts = products.slice().sort(compareProducts);
        setProducts(sortedProducts)
        setLoading(false)
      }
    });
  };

  const retryLoadDiscountedProducts = () => {
    setLoading(true)
    if (Platform.OS === "ios") {
      retryLoadDiscountediOSProducts();
    } else if (Platform.OS === "android") {
      loadAndroidDiscountedProducts();
    }
  };

  const retryLoadDiscountediOSProducts = () => {
    setLoading(true)
    InAppUtils.loadProducts(discountedIdentifiers, (error, products) => {
      if (error) {
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later"
        );
      } else if (products.length !== 2) {
        // IAP products not retrieved (App Store server down, etc.)
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later",
          [
            {
              text: "Cancel",
              style: "cancel",
            },
            {
              text: "Try Again",
              onPress: () => loadDiscountedProducts(),
            },
          ],
          { cancelable: false }
        );
      } else {
        const sortedProducts = products.slice().sort(compareProducts);
        setDiscountedProducts(sortedProducts)
        setLoading(false)
      }
    });
  };

  const loadDiscountedProducts = async () => {
    setLoading(true)
    Alert.alert("", "Discounts applied");
    if (Platform.OS === "ios") {
      loadDiscountediOSProducts();
    } else if (Platform.OS === "android") {
      loadAndroidDiscountedProducts();
    }
  };

  const loadAndroidDiscountedProducts = async () => {
    RNIap.getSubscriptions(discountedItemSku)
      .then((products) => {
        if (products.length !== 2) {
          // IAP products not retrieved (App Store server down, etc.)
          setLoading(false)
          Alert.alert(
            "Unable to connect to the Play Store",
            "Please try again later",
            [
              {
                text: "Cancel",
                style: "cancel",
              },
              {
                text: "Try Again",
                onPress: () => retryLoadDiscountedProducts(),
              },
            ],
            { cancelable: false }
          );
        } else {
          const sortedProducts = convertToProduct(products.slice().sort(compareProducts));
          setDiscountedProducts(sortedProducts)
          setLoading(false)
        }
      })
      .catch((error) => {
        setLoading(false)
        Alert.alert(
          "Unable to connect to the Play Store",
          "Please try again later"
        );
      });
  };

  const loadDiscountediOSProducts = async () => {
    await InAppUtils.loadProducts(discountedIdentifiers, (error, products) => {
      if (error) {
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later"
        );
      } else if (products.length !== 2) {
        // IAP products not retrieved (App Store server down, etc.)
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later",
          [
            {
              text: "Cancel",
              style: "cancel",
            },
            {
              text: "Try Again",
              onPress: () => retryLoadDiscountedProducts(),
            },
          ],
          { cancelable: false }
        );
      } else {
        const sortedProducts = products.slice().sort(compareProducts);
        setDiscountedProducts(sortedProducts)
        setLoading(false)      }
    });
  };

  const purchaseDiscountedProduct = async (
    productIdentifier,
    productPrice,
    productCurrencyCode
  ) => {
    setLoading(true)
    Haptics.selectionAsync();
    if (productIdentifier === undefined) {
      Alert.alert("No subscription selected");
      return;
    }
    Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
    if (Platform.OS === "ios") {
      await iOSPurchaseDiscountedProductIdentifier(
        productIdentifier,
        productPrice,
        productCurrencyCode
      );
    } else if (Platform.OS === "android") {
      RNIap.requestSubscription(productIdentifier);
    }
  };

  const loadProducts = async () => {
    setLoading(true)
    if (Platform.OS === "ios") {
      loadiOSProducts();
    } else if (Platform.OS === "android") {
      loadAndroidProducts();
    }
  };

  const loadAndroidProducts = async () => {
    try {
      RNIap.getSubscriptions(itemSkus)
        .then((products) => {
          if (products.length !== 2) {
            // IAP products not retrieved (App Store server down, etc.)
            setLoading(false)
            Alert.alert(
              "Unable to connect to the Play Store",
              "Please try again later",
              [
                {
                  text: "Cancel",
                  style: "cancel",
                },
                {
                  text: "Try Again",
                  onPress: () => retryLoadProducts(),
                },
              ],
              { cancelable: false }
            );
          } else {
            const sortedProducts = convertToProduct(
              products.slice().sort(compareProducts)
            );
            setProducts(sortedProducts)
            setLoading(false)
          }
        })
        .catch((error) => {
          setLoading(false)
          Alert.alert(
            "Unable to connect to the Play Store",
            "Please try again later"
          );
        });
    } catch (err) {
      console.warn(err); // standardized err.code and err.message available
    }
  };

  const loadiOSProducts = async () => {
    await InAppUtils.loadProducts(identifiers, (error, products) => {
      if (error) {
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later"
        );
      } else if (products.length !== 2) {
        // IAP products not retrieved (App Store server down, etc.)
        setLoading(false)
        Alert.alert(
          "Unable to connect to the App Store",
          "Please try again later",
          [
            {
              text: "Cancel",
              style: "cancel",
            },
            {
              text: "Try Again",
              onPress: () => retryLoadProducts(),
            },
          ],
          { cancelable: false }
        );
      } else {
        const sortedProducts = products.slice().sort(compareProducts);
        setProducts(sortedProducts)
        setLoading(false)
      }
    });
  };

  const handleClick = () => {
    if (!specialOffer && products) {
      purchaseProduct(
        products[selectedIndex].identifier,
        products[selectedIndex].price,
        products[selectedIndex].currencyCode
      );
    } else if (specialOffer && discountedProducts && products) {
      purchaseDiscountedProduct(
        products[selectedIndex].identifier,
        products[selectedIndex].price,
        products[selectedIndex].currencyCode
      );
    } else {
      console.log("Something went please try again...")
    }
  }

  useEffect(() => {
    navigation.setParams({ handleLogout: logout });
  }, [])

  useEffect(() => {
    async () => {
      await RNIap.initConnection();
    }
  }, [])

  useEffect(() => {
    androidSubscriptions();
  }, [])

  useEffect(() => {
    async () => {
      if (specialOffer) {
        await loadDiscountedProducts();
        await loadProducts();
      } else {
        await loadProducts();
      }
    }
  }, [])
  console.log("products: ", products)
  return (
    <React.Fragment>
      <View style={styles.container}>
        <ScrollView bounces={false} showsVerticalScrollIndicator={false}>
          <ImageBackground
            source={ONBOARDINGIMG.SUBSCRIPTIONBG}
            style={styles.imageBackgroundContainer}
          >
            <View style={styles.headerContainer}>
            </View>
          </ImageBackground>
          <View style={styles.contentContainer}>
            <Text style={styles.paymentOptionText}>
              Choose your payment option
            </Text>
            <View style={styles.subscriptionTileRow}>
              {!specialOffer &&
                products &&
                products.map((product, index) => (
                  <SubscriptionTile
                    key={product.identifier}
                    primary={
                      product.identifier.indexOf(
                        "fitazfkapp.sub.fullaccess.monthly"
                      ) > 0 ||
                      product.identifier.indexOf(
                        "fitazfkapp.sub.fullaccess.monthly.discount"
                      ) > 0
                    }
                    title={
                      Platform.OS === "ios"
                        ? productTitleMapIOS[index]
                        : productTitleMapAndroid[index]
                    }
                    price={product.priceString}
                    priceNumber={product.price}
                    currencyCode={product.currencyCode}
                    currencySymbol={product.currencySymbol}
                    onPress={() => setSelectedIndex(index)}
                    selected={selectedIndex === index}
                    term={
                      Platform.OS === "android"
                        ? andriodSubscriptionTitleMap[index]
                        : subscriptionPeriodMap[product.identifier]
                    }
                  />
                ))}
            </View>
          </View>
          <View>
            <CustomBtn
              customBtnStyle={{
                marginHorizontal: containerPadding,
                borderRadius: 50,
                marginTop: 20,
                marginBottom: 10,
              }}
              Title="Start"
              onPress={() => handleClick()}
            />
            <Icon
              name="chevron-up"
              size={8}
              color={colors.charcoal.dark}
              style={styles.chevronUp}
            />
            <View style={styles.disclaimerTextContainer}>
              <Text style={styles.disclaimerText}>
                <Text style={styles.subscriptionTermsTitle}>
                  Subscription Terms:{" "}
                </Text>
                {"By continuing, you accept our "}
                <Text
                  onPress={() =>
                    openLink(
                      "https://fitazfk.com/pages/fitazfk-app-privacy-policy"
                    )
                  }
                  style={styles.link}
                >
                  Privacy Policy
                </Text>
                {" and "}
                <Text
                  onPress={() =>
                    openLink(
                      "https://fitazfk.com/pages/fitazfk-app-terms-conditions"
                    )
                  }
                  style={styles.link}
                >
                  Terms and Conditions
                </Text>
                . You also agree that an ongoing subscription to the FitazFK
                App (FitazFK Fitness & Nutrition) will be applied to your
                iTunes account at the end of your{" "}
                {specialOffer ? "1 month" : "7 day"} free trial. Subscriptions
                will automatically renew and your account charged unless
                auto-renew is turned off at least 24-hours before the end of
                the current period. Subscriptions may be managed by the user
                and auto-renewal may be turned off by going to the users
                Account Settings after purchase. Any unused portion of a free
                trial period, if offered, will be forfeited when the user
                purchases a subscription to that publication, where
                applicable.
              </Text>
            </View>
          </View>
        </ScrollView>
      </View>
      {loading && <NativeLoader />}
    </React.Fragment>
  );
}