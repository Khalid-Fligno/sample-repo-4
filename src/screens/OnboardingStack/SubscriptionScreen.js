import React from 'react';
import {
  View,
  ScrollView,
  Text,
  StyleSheet,
  SafeAreaView,
  Dimensions,
  NativeModules,
  Alert,
  TouchableOpacity,
} from 'react-native';
import { DotIndicator } from 'react-native-indicators';
import { identifiers, validateReceiptProduction, validateReceiptSandbox, compare } from '../../../config/apple';
import SubscriptionTile from '../../components/Onboarding/SubscriptionTile';
import Loader from '../../components/Shared/Loader';
import colors from '../../styles/colors';
import fonts from '../../styles/fonts';

const { InAppUtils } = NativeModules;
const { width, height } = Dimensions.get('window');

export default class SubscriptionScreen extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      loadingProducts: false,
      loading: false,
      products: undefined,
      subscriptionSelected: undefined,
    };
  }
  componentDidMount() {
    this.props.navigation.setParams({ restore: this.restore });
    this.loadProducts();
  }
  restore = async () => {
    this.setState({ loading: true });
    InAppUtils.restorePurchases(async (error, response) => {
      if (error) {
        this.setState({ loading: false });
        Alert.alert('iTunes Error', 'Could not connect to iTunes store.');
      } else {
        if (response.length === 0) {
          this.setState({ loading: false });
          Alert.alert('No Purchases to restore');
          return;
        }
        const sortedPurchases = response.slice().sort(compare);
        try {
          const validationData = await this.validate(sortedPurchases[0].transactionReceipt);
          if (validationData === undefined) {
            Alert.alert('Receipt Validation Error');
          }
          if (validationData.latest_receipt_info && validationData.latest_receipt_info.expires_date > Date.now()) {
            this.setState({ loading: false });
            Alert.alert('Restore Successful', 'Successfully restored your purchase.');
            this.props.navigation.navigate('App');
          } else if (validationData.latest_receipt_info && validationData.latest_receipt_info.expires_date < Date.now()) {
            this.setState({ loading: false });
            Alert.alert('Expired', 'Your most recent subscription has expired');
          } else {
            this.setState({ loading: false });
            Alert.alert('No purchase information available');
          }
        } catch (err) {
          this.setState({ loading: false });
          Alert.alert('No current subscriptions to restore');
        }
      }
    });
  }
  loadProducts = () => {
    this.setState({ loadingProducts: true });
    InAppUtils.loadProducts(identifiers, (error, products) => {
      if (error) {
        Alert.alert('Could not load subscription products');
      }
      this.setState({ products, subscriptionSelected: products[1], loadingProducts: false });
    });
  }
  purchaseProduct = async (productIdentifier) => {
    this.setState({ loading: true });
    InAppUtils.canMakePayments((canMakePayments) => {
      if (!canMakePayments) {
        this.setState({ loading: false });
        Alert.alert('Not Allowed', 'This device is not allowed to make purchases. Please check restrictions on device');
      }
      InAppUtils.purchaseProduct(productIdentifier, async (error, response) => {
        if (error) {
          this.setState({ loading: false });
          Alert.alert('Something went wrong');
        }
        if (response && response.productIdentifier) {
          const validationData = await this.validate(response.transactionReceipt);
          if (validationData === undefined) {
            Alert.alert('Receipt validation error');
          }
          const isSubscribed = validationData.latest_receipt_info.expires_date > Date.now();
          if (isSubscribed === true) {
            this.setState({ loading: false });
            Alert.alert('Purchase Successful', `Your Transaction ID is ${response.transactionIdentifier}`);
            this.props.navigation.navigate('App');
          } else if (isSubscribed === false) {
            Alert.alert('Purchase Unsuccessful');
          } else {
            Alert.alert('Something went wrong', `${isSubscribed.message}`);
          }
        }
      });
    });
  }
  validate = async (receiptData) => {
    const validationData = await validateReceiptProduction(receiptData).catch(async (err) => {
      console.log(err);
      const validationDataSandbox = await validateReceiptSandbox(receiptData);
      return validationDataSandbox;
    });
    if (validationData) {
      return validationData;
    }
    return undefined;
  }
  toggleSubscriptionSelected = (subscriptionSelected) => {
    this.setState({ subscriptionSelected });
  }
  render() {
    const {
      loadingProducts,
      loading,
      products,
      subscriptionSelected,
    } = this.state;
    return (
      <SafeAreaView style={styles.container}>
        <View style={styles.flexContainer}>
          <View style={styles.headerContainer}>
            <Text style={styles.headerText}>
              Subscription
            </Text>
            <Text style={styles.subHeadingText}>
              Subscribe now for a 7-day free trial!
            </Text>
          </View>
          <View style={styles.contentContainer}>
            <View style={styles.subscriptionTileRow}>
              {
                products && products.map((product) => (
                  <SubscriptionTile
                    key={product.identifier}
                    title={product.title}
                    price={product.priceString}
                    currencyCode={product.currencyCode}
                    onPress={() => this.toggleSubscriptionSelected(product)}
                    active={subscriptionSelected === product}
                  />
                ))
              }
            </View>
          </View>
          <View style={styles.scrollViewContainer}>
            <ScrollView>
              <View style={styles.disclaimerTextContainer}>
                <Text style={styles.disclaimerText}>
                  A <Text style={{ fontFamily: fonts.bold }}>{subscriptionSelected && `${subscriptionSelected.priceString} `}{subscriptionSelected && `${subscriptionSelected.currencyCode} `}{subscriptionSelected && `${subscriptionSelected.title.toLowerCase()} `}</Text>purchase will be applied to your iTunes account at the end of your 7-day free trial.
                </Text>
                <Text style={styles.disclaimerText}>
                  Subscriptions will automatically renew unless canceled within 24-hours before the end of the current period. You can cancel anytime with your iTunes account settings. Any unused portion of a free trial will be forfeited if you purchase a subscription.
                </Text>
                <Text style={styles.disclaimerText}>
                  For more information, see our [link to ToS] and [link to Privacy Policy].
                </Text>
              </View>
            </ScrollView>
          </View>
          <View style={styles.buttonContainer}>
            <TouchableOpacity
              onPress={() => this.purchaseProduct(subscriptionSelected.identifier)}
              style={styles.button}
            >
              {
                loading ? (
                  <DotIndicator
                    color={colors.white}
                    count={3}
                    size={6}
                  />
                ) : (
                  <Text style={styles.buttonText}>
                    CONTINUE
                  </Text>
                )
              }
            </TouchableOpacity>
          </View>
          <Loader
            loading={loadingProducts}
            color={colors.coral.standard}
          />
        </View>
        {
          loading && (
            <View style={styles.overlay} />
          )
        }
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.black,
    alignItems: 'center',
  },
  flexContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: colors.offWhite,
  },
  headerContainer: {
    flexShrink: 1,
    width,
    padding: 10,
    paddingTop: 15,
  },
  headerText: {
    fontFamily: fonts.bold,
    fontSize: 28,
    color: colors.charcoal.light,
    marginBottom: 5,
  },
  subHeadingText: {
    fontFamily: fonts.standard,
    fontSize: 14,
    color: colors.charcoal.light,
    marginLeft: 3,
  },
  contentContainer: {
    flexShrink: 1,
    justifyContent: 'center',
    width,
    height: 140,
    marginBottom: 10,
  },
  scrollViewContainer: {
    flexShrink: 1,
    width: width - 20,
    backgroundColor: colors.white,
    paddingRight: 5,
    borderColor: colors.grey.light,
    borderWidth: 1,
    borderRadius: 4,
  },
  subscriptionTileRow: {
    flexDirection: 'row',
    paddingLeft: 5,
    paddingRight: 5,
  },
  disclaimerTextContainer: {
    paddingTop: 10,
    paddingLeft: 10,
    paddingRight: 5,
  },
  disclaimerText: {
    fontFamily: fonts.standard,
    fontSize: 12,
    color: colors.charcoal.light,
    marginBottom: 10,
  },
  containerButton: {
    flexShrink: 1,
    justifyContent: 'flex-end',
    height: 70,
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 50,
    margin: 10,
    width: width - 20,
    borderRadius: 2,
    shadowOpacity: 0.5,
    shadowOffset: { width: 0, height: 2 },
    shadowRadius: 3,
    backgroundColor: colors.coral.standard,
    shadowColor: colors.coral.standard,
  },
  buttonText: {
    color: colors.white,
    fontFamily: fonts.bold,
    fontSize: 16,
    marginTop: 3,
  },
  overlay: {
    position: 'absolute',
    height,
    width,
    backgroundColor: colors.black,
    opacity: 0.6,
  },
});
