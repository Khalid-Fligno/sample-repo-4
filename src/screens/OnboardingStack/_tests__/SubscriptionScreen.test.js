import React from 'react';
import renderer from 'react-test-renderer';
import SubscriptionScreen from '../SubscriptionScreen';

describe('Subscription Screen', () => {
  it('renders correctly', () => {
    const navigation = {
      getParam: jest.fn(),
    };
    const tree = renderer.create(<SubscriptionScreen navigation={navigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
