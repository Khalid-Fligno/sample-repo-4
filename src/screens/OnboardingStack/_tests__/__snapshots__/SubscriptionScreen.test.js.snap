// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Screen renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(0, 0, 0)",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    bounces={false}
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Object {
            "alignItems": "center",
            "height": 1280.6399999999999,
            "justifyContent": "space-between",
          }
        }
      >
        <Image
          source={1}
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": 1280.6399999999999,
                "width": undefined,
              },
              undefined,
            ]
          }
        />
        <View
          style={
            Object {
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(0, 0, 0)",
                "fontFamily": "GothamBold",
                "fontSize": 8,
                "textAlign": "center",
              }
            }
          >
            YOU'RE ON YOUR WAY TO
          </Text>
          <Text
            style={
              Object {
                "color": "rgb(0, 0, 0)",
                "fontFamily": "GothamBold",
                "fontSize": 8,
                "textAlign": "center",
              }
            }
          >
            GETTING SHREDDED!
          </Text>
          <Text
            style={
              Object {
                "color": "rgb(238, 48, 66)",
                "fontFamily": "GothamUltraItalic",
                "fontSize": 22,
                "textAlign": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(0, 0, 0)",
                  "fontFamily": "TuesdayNight",
                  "fontSize": 30,
                  "textAlign": "center",
                }
              }
            >
              start  
            </Text>
            YOUR 7-DAY
          </Text>
          <Text
            style={
              Object {
                "color": "rgb(238, 48, 66)",
                "fontFamily": "GothamUltraItalic",
                "fontSize": 22,
                "marginTop": -20,
                "textAlign": "center",
              }
            }
          >
            FREE TRIAL TODAY!
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "justifyContent": "flex-end",
              "width": 750,
            }
          }
        >
          <View
            style={
              Object {
                "paddingLeft": 10,
                "paddingRight": 10,
              }
            }
          />
        </View>
        <Text />
        <View
          style={
            Object {
              "padding": 20,
              "paddingTop": 10,
              "width": 750,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(255, 255, 255)",
                "fontFamily": "GothamBook",
                "fontSize": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "rgb(255, 255, 255)",
                  "fontFamily": "GothamBold",
                  "fontSize": 8,
                }
              }
            >
              Subscription Terms: 
            </Text>
            By continuing, you accept our 
            <Text
              onPress={[Function]}
              style={
                Object {
                  "color": "rgb(122, 155, 224)",
                  "textDecorationLine": "underline",
                  "textDecorationStyle": "solid",
                }
              }
            >
              Privacy Policy
            </Text>
             and 
            <Text
              onPress={[Function]}
              style={
                Object {
                  "color": "rgb(122, 155, 224)",
                  "textDecorationLine": "underline",
                  "textDecorationStyle": "solid",
                }
              }
            >
              Terms and Conditions
            </Text>
            . You also agree that an ongoing subscription to the FitazFK App (FitazFK Fitness & Nutrition) will be applied to your iTunes account at the end of your 7-day free trial. Subscriptions will automatically renew and your account charged unless auto-renew is turned off at least 24-hours before the end of the current period. Subscriptions may be managed by the user and auto-renewal may be turned off by going to the users Account Settings after purchase. Any unused portion of a free trial period, if offered, will be forfeited when the user purchases a subscription to that publication, where applicable.
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
